File: 0269-alien-dictionary/0269-alien-dictionary.py
class Solution:
    def alienOrder(self, words: List[str]) -> str:
        
        d = {}
        
        for x in words[0]:
            if x not in d:
                d[x] = []
                    
        for i in range(len(words)-1):
            w1,w2 = words[i], words[i+1]
            
            l1 = len(w1)
            l2 = len(w2)
            ml = min(len(w1),len(w2))
             
            if w1[:ml]==w2[:ml] and len(w1)>len(w2):
                return ""
            
            for x in w1:
                if x not in d:
                    d[x] = []
                
            for x in w2:
                if x not in d:
                    d[x] = []
                
            for x in range(ml):
                if w1[x]!=w2[x]:
                    d[w2[x]].append(w1[x])
                    break
            
        vis = {i:0 for i in d.keys()}
        
        res = []
        
        def iscycle(i):
            if vis[i]==1:
                return True
            
            if vis[i]==0:
                vis[i]=1
                for x in d[i]:
                    if iscycle(x):
                        return True
                    
                vis[i]=2
                
                res.append(i)
                
            return False
        
        for i in d.keys():
            if iscycle(i):
                return ""
            
        return ''.join(res)

File: 0269-alien-dictionary/0269-alien-dictionary.cpp
class Solution {
public:
    string alienOrder(vector<string>& words) {
        unordered_map<char,vector<char>> adj;
        for(int i=0;i<words.size()-1;i++){
            string w1=words[i];
            string w2=words[i+1];
            int l = min(w1.size(),w2.size());
            if(w1.substr(0,l)==w2.substr(0,l) && w1.size()>w2.size())
                return "";
            for(int i=0;i<w1.size();i++)
                if(adj.count(w1[i])==0)
                    adj[w1[i]].push_back({});
            for(int i=0;i<w2.size();i++)
                if(adj.count(w2[i])==0)
                    adj[w2[i]].push_back({});
            
            for(int i=0;i<l;i++){
                if(w1[i]!=w2[i]){
                    adj[w1[i]].push_back(w2[i]);
                    break;
                }
            }
        }
        if(words.size()==1){
            for(int i=0;i<words[0].size();i++)
                if(adj.count(words[0][i])==0)
                    adj[words[0][i]].push_back({});
        }
         
        vector<int> vis(26,0);
        string res="";
        for(auto x:adj)
            if(iscyclic(adj, x.first, res, vis))
                return "";
        
        // for(auto x:adj){
        //     cout<<x.first<<":";
        //     for(int i=0;i<x.second.size();i++)
        //         cout<<x.second[i]<<',';
        //     cout<<endl;
        // }
        reverse(begin(res),end(res));
        return res;
    }
    
    bool iscyclic(unordered_map<char,vector<char>> &adj, char c, string &res, vector<int> &vis){
        if(vis[c-'a']==1)
            return true;
        else if(vis[c-'a']==0){
            vis[c-'a']=1;
            for(char n: adj[c]){
                if(adj.count(n) && iscyclic(adj,n,res,vis))
                    return true;
            }
            res+=c;
            vis[c-'a']=2;
        }
        return false;
    }
};